### PR TITLE
Add test for volume filter function (closes #483)

### LINKED ISSUE #483: [ad-hoc] [3](5) Add test for volumeFilter to ensure we correctly handle offers that will put us over
use tests in `dailyVolumeByDate_test.go` and `TestMakeVolumeFilterConfig` as inspiration.

### PR DESCRIPTION
**Pull request by debnil**
This PR adds a test for the volume filter function (`volumeFilterFn`). It closes #483.

### CODE REVIEW

**Comment by debnil** [comment]
@nikhilsaraf: The included test cases generate all the various outcomes we care about. There are a couple edge cases that I haven't included (like testing the `=<` constraints in a couple places), because this is already a huge number of test cases. Any suggestions around readability would be great.

**Review by nikhilsaraf — COMMENTED** [review]
good idea to get the readability fixes first before expanding the number of test cases.

added suggestions to improve readability. once that's finalized, we should also add cases where the tbb values are non-zero and include larger numbers than 1.0 with more decimals. and also include numbers < 0 such as 0.123342 (or where otb + tbb = 1234.12 and capacity is 1234.52 with mode exact -- so we test that we correctly compute the difference and set the new op's amount accordingly)

File: plugins/volumeFilter_test.go [hunk 1]
```diff
@@ -145,3 +146,276 @@ func TestMakeFilterVolume(t *testing.T) {
...
+			mode:               volumeFilterModeExact,
+			sellBaseCapInBase:  pointy.Float64(0.0),
+			sellBaseCapInQuote: nil,
+			otbBaseCap:         pointy.Float64(0.0),
+			otbQuoteCap:        pointy.Float64(0.0),
+			tbbBaseCap:         pointy.Float64(0.0),
+			tbbQuoteCap:        pointy.Float64(0.0),
+			price:              "2.0",
+			inputAmount:        "100.0",
+			wantAmount:         "",
```
[the remark this answers is not in the document]
Comment (debnil) [reply plugins/volumeFilter_test.go:177]: Got it - this is what I was originally thinking, but I changed the approach because of [this previous comment](https://github.com/stellar/kelp/pull/552#discussion_r512331838). Which approach should I follow?

File: plugins/volumeFilter_test.go [hunk 2]
```diff
@@ -145,3 +146,276 @@ func TestMakeFilterVolume(t *testing.T) {
...
+
+			assert.Equal(t, wantOp, actual)
+			assert.Equal(t, wantTBBAccumulator, dailyTBBAccumulator)
+		})
+
+	}
+}
+
+func makeManageSellOffer(price, amount string) *txnbuild.ManageSellOffer {
+	if amount == "" {
```
[the remark this answers is not in the document]
Comment (debnil) [reply plugins/volumeFilter_test.go:411]: Makes sense. Given the other changes, we don't need to passthrough the empty string, so I'm just going to remove the `amount` check and early return.

File: plugins/volumeFilter_test.go [hunk 3]
```diff
@@ -145,3 +146,276 @@ func TestMakeFilterVolume(t *testing.T) {
...
+			inputAmount:        "100.0",
+			wantAmount:         "0.5000000",
+			wantTbbBaseCap:     pointy.Float64(0.5),
+			wantTbbQuoteCap:    pointy.Float64(1.0),
+		},
+		{
+			name:               "selling, quote units sell cap, keep selling quote, ignore mode",
+			mode:               volumeFilterModeIgnore,
+			sellBaseCapInBase:  nil,
+			sellBaseCapInQuote: pointy.Float64(1.0),
```
[the remark this answers is not in the document]
Comment (debnil) [reply plugins/volumeFilter_test.go:335]: I don't fully understand this comment. I get the general importance of precision, but not the rest - I've outlined specific questions below.

I think 1.0 and 0.0 are useful because there's a good amount of math in the function. For example, we add three different fields to calculate `projectedSoldInBaseUnits`. Having simple values makes it easy to debug failures and recalculate possible changes. Having arbitrary values will make that harder.

1. In the context of this function, why would some value be better for testing than another? Ones and zeros might be trivial in general, but I don't understand that in this context. Changing them to `2.34` or `5.78` doesn't test anything new. The only thing I can think about is boundary conditions, but that's separate from the triviality of values.
2. "We actively modify the numbers of offers" - where is this happening? I might be missing something here entirely, but this seems outside the scope of the function we are testing.
3. How would this function affect the precision of anything? Precision could matter where `opToReturn.Amount` is defined. But there, the `fmt.Sprintf` guarantees that `newAmountBeingSold` is converted to a string with 7 decimal places of precision. And the rest of the operations are just addition.

I'm definitely not opposed to changing these values, but I do want to choose meaningful values that increase test coverage, rather than making test failures harder to debug.

File: plugins/volumeFilter_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nikhilsaraf) [reply plugins/volumeFilter_test.go]: we should follow the approach suggested in this PR, thanks for checking 👍

File: plugins/volumeFilter_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikhilsaraf) [reply plugins/volumeFilter_test.go:323]: 1. the reason I suggest not using 0.0 and 1.0 is because 0.0 is a nop for addition and 1.0 is a nop for multiplication. so in general those numbers should be avoided when testing math based functions. instead we can use values such as 0.5 and 1.5 for example. 

as an example, if our math in the function does the equivalent of this:
```
add(a, b) {
    return a + b + b
}
```
then `assertEqual(1.0, add(1.0, 0.0))` will be `true` even though the function is broken.
so it's ok to have this in conjunction with another test case such as `assertEqual(1.5, add(1.0, 0.5))`, which will correctly fail. Similarly for multiply when using the value `1.0`

2. sorry, I meant "numbers in the offers".

3. since the final value used in manage offer is a string representation and we are testing the op, then we will automatically test the precision as well. what i meant was that if we test something like `1.001 * 0.12105`, what is the result from rounding, `0.1211710` (round down) or `0.1211711` (round up)?
the test should capture it whatever it is. Then we need to think about what do we want it to be, round up or round down? then we modify the code accordingly and update the test. we can have just 1-2 special test cases for this calling them the "rounding" checks so the remaining tests use simpler numbers.

File: plugins/volumeFilter_test.go [hunk 4]
```diff
@@ -145,3 +146,229 @@ func TestMakeFilterVolume(t *testing.T) {
...
+			otbQuote:           pointy.Float64(0.0),
+			tbbBase:            pointy.Float64(0.0),
+			tbbQuote:           pointy.Float64(0.0),
+			inputOp:            makeManageSellOffer("2.0", "100.0"),
+			wantOp:             nil,
+			wantTbbBase:        pointy.Float64(0.0),
+			wantTbbQuote:       pointy.Float64(0.0),
+		},
+		{
+			name:               "9. selling, quote units sell cap, keep selling quote, new amount, exact mode",
```
Comment (nikhilsaraf) [inline plugins/volumeFilter_test.go:278]: seems same as # 11 below?
what do you mean by "new amount" in the name? -- is that a remnant of the old set of test cases?

File: plugins/volumeFilter_test.go [hunk 5]
```diff
@@ -145,3 +146,229 @@ func TestMakeFilterVolume(t *testing.T) {
...
+			otbQuote:           pointy.Float64(0.0),
+			tbbBase:            pointy.Float64(0.0),
+			tbbQuote:           pointy.Float64(0.0),
+			inputOp:            makeManageSellOffer("2.0", "100.0"),
+			wantOp:             makeManageSellOffer("2.0", "0.5000000"),
+			wantTbbBase:        pointy.Float64(0.5),
+			wantTbbQuote:       pointy.Float64(1.0),
+		},
+		{
+			name:               "10. selling, quote units sell cap, keep selling quote, new amount, ignore mode",
```
Comment (nikhilsaraf) [inline plugins/volumeFilter_test.go:292]: similarly, seem like a repeat of # 12 below

File: plugins/volumeFilter_test.go [hunk 6]
```diff
@@ -145,3 +146,229 @@ func TestMakeFilterVolume(t *testing.T) {
...
+	}
+
+	// we fix the marketIDs and accountIDs, since volumeFilterFn output does not depend on them
+	marketIDs := []string{}
+	accountIDs := []string{}
+
+	for _, k := range testCases {
+		t.Run(k.name, func(t *testing.T) {
+			// exactly one of the two cap values must be set
+			if assert.Nil(t, k.sellBaseCapInBase) == assert.Nil(t, k.sellBaseCapInQuote) {
```
Comment (nikhilsaraf) [inline plugins/volumeFilter_test.go:342]: when you run this, if one of them is non-nil then it will produce a test failure no? (because we are using assert to get the boolean value)

for a good way to achieve what we want without frivolous errors we can try this:
```golang
if k.sellBaseCapInBase == nil && k.sellbaseCapInQuote == nil {
    assert.Fail(t, "either one of the two cap values must be set")
    return
}
if k.sellBaseCapInBase != nil && k.sellbaseCapInQuote != nil {
    assert.Fail(t, "both of the cap values cannot be set")
    return
}
```

File: plugins/volumeFilter_test.go [hunk 7]
```diff
@@ -145,3 +146,229 @@ func TestMakeFilterVolume(t *testing.T) {
...
+			assert.Equal(t, k.wantOp, actual)
+
+			wantTBBAccumulator := makeRawVolumeFilterConfig(k.wantTbbBase, k.wantTbbQuote, k.mode, marketIDs, accountIDs)
+			assert.Equal(t, wantTBBAccumulator, dailyTBBAccumulator)
+		})
+
+	}
+}
+
+func makeManageSellOffer(price, amount string) *txnbuild.ManageSellOffer {
```
Comment (nikhilsaraf) [inline plugins/volumeFilter_test.go:367]: `(price string, amount string)`

**Review by nikhilsaraf — COMMENTED** [review]
added comments. I think we need some more test cases where the tbb values are not 0 as well. Currently it seems that if we removed the tbb reading logic from the code the tests would still pass.

File: plugins/volumeFilter_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (debnil) [reply plugins/volumeFilter_test.go:323]: Awesome, thanks for the super thorough reply! This is a really helpful framework for thinking about this. Agreed that we can have a couple tests for the rounding checks.

File: plugins/volumeFilter_test.go (hunk 6 above)
Comment (debnil) [reply plugins/volumeFilter_test.go:342]: Ah, this makes sense, thanks!

**Review by nikhilsaraf — APPROVED** [review]
lgtm, thanks! 🚢
